src/main/ruby/core/string.rb

```diff
@@ -643,21 +643,32 @@ def upto(stop, exclusive=false)
     stop = StringValue(stop)
 
     if stop.size == 1 && size == 1
+      enc = Encoding.compatible?(self.encoding, stop.encoding)
+
       return self if self > stop
       after_stop = stop.getbyte(0) + (exclusive ? 0 : 1)
       current = getbyte(0)
```

Comment:
This seems to assume there is only a single byte, but `size` is used, not `bytesize`.